plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/pages/utils.kt

```diff
@@ -1,15 +1,22 @@
 package org.jetbrains.dokka.javadoc.pages
 
 import org.jetbrains.dokka.model.*
+import org.jetbrains.dokka.utilities.htmlEscape
 
-internal fun JavadocFunctionNode.getAnchor(): String =
-    "$name(${parameters.joinToString(",") {
-        when (val bound = if (it.typeBound is Nullable) it.typeBound.inner else it.typeBound) {
-            is TypeConstructor -> listOf(bound.dri.packageName, bound.dri.classNames).joinToString(".")
-            is TypeParameter -> bound.name
-            is PrimitiveJavaType -> bound.name
-            is UnresolvedBound -> bound.name
-            is JavaObject -> "Object"
-            else -> bound.toString()
-        }
-    }})"
\ No newline at end of file
+internal fun JavadocFunctionNode.getAnchor(): String {
+    val parameters = parameters.joinToString(",") { it.typeBound.asString() }
+    return "$name($parameters)".htmlEscape()
```

Comment:
Why is `htmlEscape` needed here? `UrlEncode`?

Reply:
> Why is htmlEscape needed here? 

The returned value is used in the HTML template directly, so if `toString()` returned something like `trololo" something else`, it would result in `<a href="trololo" something else>`, allowing for easy html injections and in general invalid HTML. Added a comment with a brief explanation

Look at the screenshots in #2803 for more examples/context .

> UrlEncode?

I'm afraid `UrlEncode` cannot be used on the whole anchor because it has stricter rules and will not allow `(`, `)`, `<` and `>` - all of which are used by the Javadoc to generate links. See [AbstractList.html#%3Cinit%3E()](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/AbstractList.html#%3Cinit%3E()) for instance

I could use `UrlEncode` selectively instead of using it on the resulting string, but I don't think it's worth the effort as there's no constant threat, it's more or less "just in case" that doesn't cost us much

Reply:
For your example:
`<init>`.htmlEscape() returns `&lt;<init>&gt;` (it brokes a link).
but `URLEncode` returns `%3Cinit%3E`. 
On [AbstractList.html#%3Cinit%3E()](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/AbstractList.html#%3Cinit%3E())  we have `<a href="#%3Cinit%3E()" class="member-name-link">AbstractList</a>`.

Reply:
Agree here with @vmishenev.

Sorry for the long-winded text below, hope this helps in some way to move this along.

Note that the output code is more a part of URL than part of HTML. HTML escape is (by the name) is meant for outputting text as HTML so HTML language-special characters need to be escaped. That said URL escape is probably not good either because that's meant for **referencing** an ID. So if you have `id="<init>()"` you need `href="#%3Cinit%3E()`, but if you URL escape the ID: `id="%3Cinit%3E()`, you'll need `href="#%253Cinit%253E()"`.

So this pretty much means that you cannot just use the same string for id/name and href, as they need different escapes.

Note that HTML was very restrictive on IDs: https://www.w3.org/TR/html4/types.html#:~:text=ID%20and%20NAME%20tokens%20must,periods%20(%22.%22), while HTML 5 is permissive: https://stackoverflow.com/a/31773673/253468

That said, the overall HTML code has to be valid (parseable). As far as I know, to achieve this, you really only need to escape the `"` with `&quot;` (as to not accidentally terminate the attribute), and the rest can be verbatim.

Hrefs are another matter, because it doesn't only need to be valid HTML, but the parsed value must be a valid URL (RFC for it is quite complex) and it actually should point to a real location. So in this case one would need to URL encode the link (percent encoding), and then on top of that make sure that it can be safely put into an HTML attribute (use correct XML/HTML entities). Note that url-encoding too much is usually not a problem, the browser will just decode each percent-escape; however double encoding is a usual problem that the browser cannot resolve, so us devs have to be careful.

This complexity will stay regardless of how you render HTML, but different templating engines make it easier or harder to do this.

Reply:
As I've mentioned, I don't think I can encode the whole anchor. The example with `<init>()` demonstrates that Javadoc applies encoding selectively. It encodes `<init>`, but does not encode `()`. Moreover, when signature links contain multiple arguments, it does not encode commas (for example, `addAll(int,java.util.Collection)`).

If I encode the whole anchor, it would be `%3Cinit%3E%28%29` and `addAll%28int%2Cjava.util.Collection%29`, which might work, but won't look as good and might lead to problems - I'm not sure.

As I said, I could apply encoding selectively (i.e encode every param separately), but you can see all possible param values just underneath - I don't see any that need encoding, and Dokka does not actually use `<init>()` and instead uses `ClassName()` for referencing constructors.

> HTML escape is (by the name) is meant for outputting text as HTML so HTML language-special characters need to be escaped. 

Unfortunately, that's what it's doing - whatever string `getAnchor()` generates, it'll be put into the HTML as is, without any prior escaping: `<a id="{{ method.anchorLink }}">`. 

That's the primary reason I put the escape there - to avoid possible invalid HTML, and to warn future generations and make them think about their changes to this function.

> So this pretty much means that you cannot just use the same string for id/name and href, as they need different escapes.

That's actually a really good point that I didn't even think about - this function is indeed used for generating both the id and the href value, which is not correct. Javadoc doesn't actually encode IDs: `<section .. id="<init>()">`

____

To sum up,

* URL encoding the anchor seems to break IDs and/or links
* HTML escaping the anchor doesn't seem to break anything yet, but it will if/when `<init>` gets introduced
* `getAnchor` needs to be refactored into two separate functions, one for generating the id and another one for the link, each with different escaping
* At the moment, unless names permit, nothing actually needs encoding/escaping

This issue/PR is about embarrassing configuration details that leak in the generated HTML, not about encoding/escaping, and I'd like to keep it short and just about that. This part has been fixed by removing the `else` branch for the sealed `Bound`, and by properly handling typealiases, which caused the problem.

The `htmlEscape` part is just extra that I added to bring attention to the problem and really just in case. I'm happy with removing it and just leaving a detailed comment with a link to this discussion, or we can leave it be and I'll still add the comment.

Thoughts?

Reply:
> If I encode the whole anchor, it would be %3Cinit%3E%28%29 and addAll%28int%2Cjava.util.Collection%29, which might work, but won't look as good and might lead to problems - I'm not sure.

What problems? 

> Javadoc doesn't actually encode IDs: <section .. id="<init>()">

It encodes
`<section class="detail" id="&lt;init&gt;()">`

> getAnchor needs to be refactored into two separate functions, one for generating the id and another one for the link, each with different escaping

`getAnchor` should not depend on the output format or the place of use.
Anyway, it does not make sense to have `htmlEscape` here.

Reply:
I vote for removing the htmlEscape from this PR and merging it. And we can open another issue where the escaping is worked out without hacking around. Just in case adding a function where it doesn't belong just creates tech debt and breaks things without knowing.

The model should never be aware of where the output goes. Just generate the preferred name and that's it. It is then the up to whoever builds the HTML to get it right:
```html
<a id="{{ HTMLEscape(method.anchorLink) }}">
<a href="{{ HTMLEscape(URLEncode(method.anchorLink)) }}">
```
if you use a half-decent templating library this should be done for you... `{{}}` looks like Mustache, but the word doesn't exist in the codebase, what is used to render HTML?

> won't look as good

I think it's better to be functional and "just works" than looking good :)
Did you consider the case that someone will copy those links and if it's too eagerly not-escaped it might not work in GitHub / Slack / GMail etc. I'm thinking about https://host.name/path/to/resource#whops a-space_in(the,url) vs https://host.name/path/to/resource#whops%20a-space_in(the,url) . The first one is borked, the second one is ugly, but at least it's clickable.

Reply:
> It encodes <section class="detail" id="&lt;init&gt;()">

It html escapes the IDs, it doesn't URLEncode them, as far as I can see. Otherwise you would not have `()` in there, it would be `%28%29`. As @TWiStErRob has pointed out, URLEncoding the id would also require a different href.

> getAnchor should not depend on the output format or the place of use.

> The model should never be aware of where the output goes. Just generate the preferred name and that's it. It is then the up to whoever builds the HTML to get it right:
> 
> <a id="{{ HTMLEscape(method.anchorLink) }}">
> <a href="{{ HTMLEscape(URLEncode(method.anchorLink)) }}">

> I think it's better to be functional and "just works" than looking good :)

In an ideal world where we develop our own format - I would totally agree with you. Here, however, Dokka tries to clone the Javadoc format, which does not encode parentheses and commas - you seem to be ignoring this fact :) 

I guarantee you, if we don't follow it, someone will be comparing the links and will create a bug report, asking "Why, for the same input, does the Javadoc anchor link have plain symbols in `addAll(int,java.util.Collection)`, and Dokka's Javadoc anchor link contains gibberish in `addAll%28int%2Cjava.util.Collection%29`". 

I'm also not sure how the tools that accept Javadoc HTML as input will react (like IntelliJ for `javadoc.jar`), if they fully decode the URL or not - the kind of possible problems I was referring to.

If we want to honor the Javadoc format rules, right now I don't see any other way but to have two different implementations that apply `URLEncode`  selectively, ignoring parantheses and commas in the generated anchor itself.

Out of curiosity I had a look at how the actual Javadoc tool does it, and they just have their own implementation for `encodeURL` that is less restrictive: https://github.com/openjdk/jdk/blob/b88ee1ee22a4ea859f2a7bdf80a12c1d56fe6fd2/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlTree.java#L299-L325

^ This would work and be neat, we won't need two different implementations then. But if we want to use the standard `URLEncoder` for the **whole** anchor, it'll lead to problems, imho

___

I've removed `htmlEscape` and left a comment. Hopefully, our discussion will help whoever fixes it in the future.

Reply:
> Dokka tries to clone the Javadoc format, which does not encode parentheses and commas - you seem to be ignoring this fact :)

I wasn't aware this is a strict requirement. Also keep in mind that you cannot do that 100%, because Kotlin has free-form names, while Java is pretty strict with naming.

> they just have their own implementation

They are aware of what the possible characters are, so they wrote their own. You'll need to do the same, but beware of ``fun `this (is) "a" test!#%20$`() { }`` and similar valid Kotlin names. I'm pretty certain that if I document this name, and generate Dokka Javadoc, it just won't work.